src/lib/structures/base/Store.ts

```diff
@@ -0,0 +1,192 @@
+
+	/**
+	 * The core directories pieces of this store can hold.
+	 * @since 0.0.1
+	 */
+	private readonly coreDirectories = new Set<string>();
+
+	/**
+	 * @since 0.0.1
+	 * @param client The client this Store was created with
+	 * @param name The name of this store
+	 * @param holds The type of structure this store holds
+	 */
+	public constructor(client: Client, name: string, holds: PieceConstructor<V>) {
+		super();
+
+		this.client = client;
+		this.name = name;
+		this.holds = holds;
+	}
+
+	/**
+	 * The directory of local pieces relative to where you run Klasa from.
+	 * @since 0.0.1
+	 */
+	public get userDirectory(): string {
+		return join(this.client.userBaseDirectory, this.name);
+	}
+
+	/**
+	 * Registers a core directory to check for pieces.
+	 * @since 0.0.1
+	 * @param directory The directory to check for core pieces
+	 */
+	public registerCoreDirectory(directory: string): this {
+		this.coreDirectories.add(join(directory, this.name));
+		return this;
+	}
+
+	/**
+	 * Initializes all pieces in this store.
+	 * @since 0.0.1
+	 */
+	public init(): Promise<Array<any>> {
+		return Promise.all(this.map(piece => piece.enabled ? piece.init() : piece.unload()));
+	}
+
+	/**
+	 * Loads a piece into Klasa so it can be saved in this store.
+	 * @since 0.0.1
+	 * @param directory The directory the file is located in
+	 * @param file A string or array of strings showing where the file is located.
+	 */
+	public async load(directory: string, file: readonly string[]): Promise<V | null> {
+		const loc = join(directory, ...file);
+		let piece = null;
+		try {
+			const loaded = await import(loc) as { default: PieceConstructor<V> } | PieceConstructor<V>;
+			const LoadedPiece = 'default' in loaded ? loaded.default : loaded;
+			if (!isClass(LoadedPiece)) throw new TypeError('The exported structure is not a class.');
+			piece = this.set(new LoadedPiece(this, directory, file));
+		} catch (error) {
+			this.client.emit('wtf', `Failed to load file '${loc}'. Error:\n${error.stack || error}`);
+		}
+		delete require.cache[loc];
+		module.children.pop();
+		return piece;
+	}
+
+	/**
+	 * Loads all of our Pieces from both the user and core directories.
+	 * @since 0.0.1
+	 * @returns The number of Pieces loaded.
+	 */
+	public async loadAll(): Promise<number> {
+		this.clear();
+		if (!this.client.options.disabledCorePieces.includes(this.name)) {
+			for (const directory of this.coreDirectories) await Store.walk(this, directory);
+		}
+		await Store.walk(this);
+		return this.size;
+	}
+
+	/**
+	 * Sets up a piece in our store.
+	 * @since 0.0.1
+	 * @param piece The piece we are setting up
+	 */
+	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+	// @ts-ignore
+	public set(piece: V): V | null {
+		if (!(piece instanceof this.holds)) {
+			this.client.emit('error', `Only ${this} may be stored in this Store.`);
+			return null;
+		}
+
+		const existing = this.get(piece.name);
+		if (existing) this.delete(existing);
```

Comment:
Any reason for get/delete? I mean, delete checks itself if the piece existed before
```suggestion
		this.delete(piece);
```